File: backend/src/cli/parse.ts
import type { CliOverrides } from './options'
import { defaultCliOverrides } from './options'

export interface ParsedCli {
  overrides: CliOverrides
}

function readLongFlagValue(argv: string[], i: number, name: string): { value: string, nextIndex: number } {
  const arg = argv[i]!
  const prefix = `${name}=`
  if (arg.startsWith(prefix)) {
    return { value: arg.slice(prefix.length), nextIndex: i + 1 }
  }
  const value = argv[i + 1]
  if (!value || value.startsWith('-')) {
    throw new Error(`Missing value for ${name}`)
  }
  return { value, nextIndex: i + 2 }
}

/** Supports `-p 3100` and `-p3100`. */
function readShortFlagValue(argv: string[], i: number, flag: string): { value: string, nextIndex: number } {
  const arg = argv[i]!
  if (arg.length > flag.length) {
    return { value: arg.slice(flag.length), nextIndex: i + 1 }
  }
  const value = argv[i + 1]
  if (!value || value.startsWith('-')) {
    throw new Error(`Missing value for ${flag}`)
  }
  return { value, nextIndex: i + 2 }
}

export function parseArgv(argv: string[]): ParsedCli {
  const overrides = defaultCliOverrides()
  let i = 0

  while (i < argv.length) {
    const arg = argv[i]!

    if (arg === 'help' || arg === '--help' || arg === '-h') {
      overrides.help = true
      i++
      continue
    }
    if (arg === '--version' || arg === '-v') {
      overrides.version = true
      i++
      continue
    }
    if (arg === '--no-tui') {
      overrides.noTui = true
      i++
      continue
    }
    if (arg === '--create-config') {
      overrides.createConfig = true
      i++
      continue
    }
    if (arg === '--with-tls') {
      overrides.withTls = true
      i++
      continue
    }
    if (arg === '--port' || arg.startsWith('--port=')) {
      const { value, nextIndex } = readLongFlagValue(argv, i, '--port')
      overrides.port = value
      i = nextIndex
      continue
    }
    if (arg === '-p' || (arg.startsWith('-p') && arg.length > 2)) {
      const { value, nextIndex } = readShortFlagValue(argv, i, '-p')
      overrides.port = value
      i = nextIndex
      continue
    }
    if (arg === '--host' || arg.startsWith('--host=')) {
      const { value, nextIndex } = readLongFlagValue(argv, i, '--host')
      overrides.host = value
      i = nextIndex
      continue
    }
    if (arg === '-H' || (arg.startsWith('-H') && arg.length > 2)) {
      const { value, nextIndex } = readShortFlagValue(argv, i, '-H')
      overrides.host = value
      i = nextIndex
      continue
    }
    if (arg === '--data-dir' || arg.startsWith('--data-dir=')) {
      const { value, nextIndex } = readLongFlagValue(argv, i, '--data-dir')
      overrides.dataDir = value
      i = nextIndex
      continue
    }

    throw new Error(`Unknown argument: ${arg}`)
  }

  return { overrides }
}

File: backend-go/cli/parse.go
package cli

import (
	"errors"
	"flag"
	"fmt"
	"io"
	"strings"
	"unicode"

	"file-lite-go/config"
)

// expandGluedShortFlags turns -p3100 / -H0.0.0.0 into -p 3100 / -H 0.0.0.0.
func expandGluedShortFlags(args []string) []string {
	out := make([]string, 0, len(args))
	for _, arg := range args {
		if strings.HasPrefix(arg, "-p") && len(arg) > 2 && unicode.IsDigit(rune(arg[2])) {
			out = append(out, "-p", arg[2:])
			continue
		}
		if strings.HasPrefix(arg, "-H") && len(arg) > 2 && !strings.HasPrefix(arg, "--") {
			out = append(out, "-H", arg[2:])
			continue
		}
		out = append(out, arg)
	}
	return out
}

func ParseArgv(args []string) (Overrides, error) {
	var o Overrides
	fs := flag.NewFlagSet(config.PkgName, flag.ContinueOnError)
	fs.SetOutput(io.Discard)

	fs.BoolVar(&o.Help, "help", false, "Show help")
	fs.BoolVar(&o.Help, "h", false, "Show help")
	fs.BoolVar(&o.Version, "version", false, "Show version")
	fs.BoolVar(&o.Version, "v", false, "Show version")
	fs.BoolVar(&o.NoTui, "no-tui", false, "Run without interactive menu")
	fs.BoolVar(&o.CreateConfig, "create-config", false, "Create config.json if missing")
	fs.BoolVar(&o.WithTLS, "with-tls", false, "With --create-config: generate self-signed cert")
	fs.StringVar(&o.Port, "port", "", "Override listen port")
	fs.StringVar(&o.Port, "p", "", "Override listen port")
	fs.StringVar(&o.Host, "host", "", "Override listen host")
	fs.StringVar(&o.Host, "H", "", "Override listen host")
	fs.StringVar(&o.DataDir, "data-dir", "", "Data directory")

	expanded := expandGluedShortFlags(args)
	if err := fs.Parse(expanded); err != nil {
		if errors.Is(err, flag.ErrHelp) {
			o.Help = true
			return o, nil
		}
		return o, err
	}
	if rest := fs.Args(); len(rest) > 0 {
		return o, fmt.Errorf("unknown argument: %s", rest[0])
	}
	return o, nil
}
